### DISCUSSION TITLE
A deep dive into the Elixir AST: Building a typed struct macro

### DISCUSSION THREAD

**Discussion by giscus** [doorgan/blog#9] [General]
# A deep dive into the Elixir AST: Building a typed struct macro

Part 2 of The Elixir AST series. Analyzing the AST to build a typed struct macro.

https://dorgan.ar/posts/2021/04/the_elixir_ast_typedstruct/

**Comment by alaadahmed** [comment]
Thank you very much for your deep tutorial, I really enjoyed it and learned a lot. Unfortunately code didn't work for me and produce error
```
== Compilation error in file lib/report.ex ==
** (FunctionClauseError) no function clause matching in TypedStruct.get_field_data/1    
    
    The following arguments were given to TypedStruct.get_field_data/1:
    
        # 1
        {:field, [line: 7], [:type, {:report_type, [line: 7], nil}, [required?: true, default: :refactoring]]}
    
    lib/typed_struct.ex:42: TypedStruct.get_field_data/1
    (elixir 1.13.0) lib/enum.ex:1593: Enum."-map/2-lists^map/1-0-"/2
    expanding macro: TypedStruct.typedstruct/1
    lib/report.ex:6: Report (module)
```
I think because in the function `get_field_data/1` you pattern match empty list to non-empty list and this will cause compilation error, I made a pull request, passing a variable to match what in `metadata` of the node will make code work.

**Reply by doorgan** [reply]
I'm glad this was useful! And thanks for the fix PR :)

**Comment by altuntasfatih** [comment]
Thank you for the excellent tutorial.
Just a quick point, To decide whether the field is enforced, it gets the ": enforced?" option. But in the field definition, you have put the ": required?" option so that reason macro parse all fields as non-enforced. Due to that, one of them should be changed.
